### ISSUE TITLE
Regression: Job dependency not respected in 0.7.0?

### ISSUE THREAD

**Issue by WeatherGod** [nvie/rq#762]
Just tried out rq-0.7.0 from pip, and while job dependencies were being registered (I manually inspected them in the redis server), the second job executed before the first job completed. Reverting back to 0.6.0 fixed this problem.

Some output from my redis instance:
```
>>> from redis import Redis
>>> from rq import Queue
>>> redis_conn = Redis()
>>> q = Queue('radarOps', connection=redis_conn)
>>> j = q.fetch_job('8')
>>> j
Job('8', enqueued_at=datetime.datetime(2016, 11, 18, 20, 49, 20))
>>> j2 = q.fetch_job('9')
>>> j2
Job('9', enqueued_at=datetime.datetime(2016, 11, 18, 20, 49, 20))
>>> j2.dependency
Job(u'8', enqueued_at=datetime.datetime(2016, 11, 18, 20, 49, 20))
>>> j2.started_at
datetime.datetime(2016, 11, 18, 20, 49, 20)
>>> j.ended_at
datetime.datetime(2016, 11, 18, 20, 49, 38)
>>>
```
So, you see that job 8 ended at 20:49:38, but job 9, which depends on job 8 got started at 20:49:20, which is the same time it was enqueued.

One thing that may or may not be relevant here is that I have ttl and result_ttl set to -1 for these jobs.

**Comment by selwin** [comment]
Jobs that depend on other jobs to finish shouldn't be placed on Redis queue if their parent jobs didn't complete successfully. 

Could you please provide a snippet of how you enqueue these jobs? Better still, make a failing test case for this.

**Comment by WeatherGod** [comment]
When you say "shouldn't be placed on Redis queue", do you mean that I
shouldn't enqueue() the second job until I see that the first job is
finished, or are you just stating the expected behavior that the worker
shouldn't be consuming that second enqueued job until the first one
completed (which I totally agree it shouldn't).

I am currently on vacation for the week and have limited time to construct
an example at the moment. I'll at least explain what I am doing, though. I
am creating an SGE clone using rq. So I have "qsub" commands that are being
called, each one submitting a job to the queue. In this particular case, I
have a script that qsub'ed the first job, got its job id, and then qsub'ed
the second job with the first job (fetched from the queue using the passed
in id) as an argument to depends_on for the second job.

I am new to redis-queue, so perhaps how I am treating queues isn't quite
right?

**Comment by WeatherGod** [comment]
Still investigating this further. This might not be a regression, per se. Instead, there might be a race condition that is made more likely in 0.7.0.  I am finding that with 0.6.0 on my test environment, I am sometimes getting jobs fired off simultaneously (bad), and sometimes the job waits like it should.

**Comment by selwin** [comment]
Could you please try running your tests with this branch and see if the race condition has been fixed? https://github.com/nvie/rq/pull/742

**Comment by selwin** [comment]
Actually, I just merged the branch into master. You could tr it in the current master branch to see if it solves the race condition problem

**Comment by WeatherGod** [comment]
Just gave it a whirl, and the problem does not seem to be fixed yet. The unit tests pass on my system, though.

I am still trying to rule out environment differences. When dependent jobs are initiated by scripts called from the crontab, that's when it seems that most dependent jobs are executed prior to the completion of the dependency job. However, if I call the same script with the same arguments from the command-line, it works exactly as expected most, if not all, the time.

**Comment by WeatherGod** [comment]
(btw, the unit tests left behind a new queue called "queue_name" and a uncompleted job in that queue; shall I open a new issue for that?)

**Comment by WeatherGod** [comment]
I have been reading through the tests in test_worker.py, and I think I can see a subtle problem with some of the tests. particularly, the test "test_job_dependency()" has two short-comings. First, it merely tests that the second job was executed, it doesn't test if the second job was started *after* the first one finished. One way to test that is to have the first test generate a file, and then have the second test read that file back in.

Second, if I am understanding the test suite correctly, only one worker is used for a queue at any given time. This would mask dependency handling issues, as funnelling jobs through a single worker would, in effect, force dependency ordering. Is there a way in the test suite to initiate the work loop of multiple workers prior to enqueueing any jobs?

**Comment by WeatherGod** [comment]
I have managed to create a test class that creates some Worker processes in its setUp() and terminate() them in the tearDown() stage, and then have it run a test of the dependency handling. I am still trying to get it to trigger my bug, though.

I can post a branch of my work so far, if anybody is interested.

**Comment by selwin** [comment]
@WeatherGod we don't use two workers to test the job dependency because dependent jobs aren't yet in `Queue` (and thus shouldn't be executed by workers).

It would be good if you can create a test case that highlights this issue. Having said that, are you sure that the clocks on your workers in sync?

**Comment by WeatherGod** [comment]
@selwin, I am wondering if we are talking past each other? I'll restate what I am experiencing. I have a Redis server, a queue and 12 workers attached to that queue (all of this is completely local, all owned and initiated by a single user -- no multi-machine or anything fancy). In a shell loop, I'll submit job A for parameters X_i, get that job's ID, then submit job B for parameters X_i, and indicate that it is dependent upon the job ID that I just got. This'll loop about 15 times, creating 15 pairs of dependent jobs.

Job A takes approximately 2-5 minutes to complete, Job B takes about 30 seconds to complete. The for-loop that submitted these jobs only takes a few seconds to complete. Therefore, while some jobs are immediately started (12 of them), there are job pairs that lie in wait (this is an important fact in a moment).

What I am experiencing is that Job B_i is *started* at the same time (or very nearly) as Job A_i. Job B_i then fails because it can't find the file that is supposed to be produced by Job A_i. That file does eventually get made, but Job B_i has long since finished. What is very interesting is that this is true regardless of whether the queue is empty or has many jobs already.

Creating a SSCCE for this has proved difficult, partly because I am still learning `rq`, and partly because I haven't yet figured out why this bug is repeatable for me when the scripts are executed from the cron, but everything is fine when executed manually.

As for the single worker question in the test suite, perhaps I am not understanding what is going on. But, if there is only ever one worker, then it is impossible to test for some sort of bug where two or more workers may accidentally consume jobs from the queue that should still be waiting?

Today, I am going to continue to try an make an SSCCE. I am also going to see what happens on my system if I reduce the number of workers from 12 to 1. If my understanding of everything is correct, then this should "fix" the problem by preventing Job B_i from starting at the same time as Job A_i.

**Comment by WeatherGod** [comment]
hmmm, a possible pattern I am seeing as a difference between the crontab jobs and manual execution. Job A_i and Job B_i, when submitted via the script from crontab have the same enqueue time down to the second, while typically they are different by a second when submitted via the same script, but manually. Could there be a hashing problem? Maybe something with regards to uuids? Anything I could look for in the redis server to verify this?

**Comment by WeatherGod** [comment]
... and that pattern doesn't seem to hold out. drats...

I can confirm, though, that reducing the number of workers down to 1 does make the jobs happen in series. So, it does look like the test suite could be missing an important edge case.

**Comment by WeatherGod** [comment]
I am feeling really stupid... The difference between the manual runs and the crontab runs was that I was doing the manual ones *after* the crontab attempt. Job A_i from the crontab run had left its artifact which Job B_i from the manual execution then picked up, making it seem like everything was working when executed manually!

I am now working from a fresh environment. It is now seeming like I was picking up the released rq-0.7.0 as installed by pip, rather than the patched 0.7.0, which was also causing confusion. I will now be able to see if the race condition patch actually fixed things. I'll also see if the test that I developed actually fails on the released 0.7.0.

btw, have you considered utilizing versioneer (https://github.com/warner/python-versioneer)? It can help situations like this by making sure that different development versions are distinguishible. The matplotlib project (which I am a member of) has had tremendous success using it.

**Comment by WeatherGod** [comment]
Progress!

So, once I was able to sort out all of the other red herrings, I was able to drill down to how I had implemented my "qsub" command as a possible source of the problem. Essentially, I had done this:
```
...
hold_job = Job.fetch(args.hold_jids) if args.hold_jids else None
job = Job.create(...., depends_on=hold_job, ...)
q.enqueue_job(job)
...
```
I had done this to get around connection errors I was having with `q.enqueue_call()`, not realizing that it isn't equivalent. Still working on confirming the last little details, but it might seem to be mostly figured out.

There are still some lingering issues:
1) clearer documentation `q.enqueue_job()`? also, should a job that has a dependency still get pushed forward by `q.enqueue_job()` even if the dependency is still running?
2) `q.enqueue_call()` has a possible bug where if a Connection is not pushed onto a stack via a context manager, then the call to `q.enqueue_call(..., depends_on=hold_job)` will fail due to line 212 in queue.py: `if not self.job_class.exists(depends_on.id):`, because it isn't given a connection object to use.
